BinaryEncoding.md

```diff
@@ -92,42 +95,46 @@ for all sections. The encoding of all sections begins as follows:
 Each section is optional and may appear at most once.
 Known sections (from this list) may not appear out of order.
 
-* [Signatures](#signatures-section) section
-* [Import Table](#import-table-section) section
-* [Function Signatures](#function-signatures-section) section
-* [Indirect Function Table](#indirect-function-table-section) section
+* [Type](#type-section) section
+* [Import](#import-section) section
+* [Function](#function-section) section
+* [Indirection](#indirection-section) section
 * [Memory](#memory-section) section
-* [Export Table](#export-table-section) section
-* [Start Function](#start-function-section) section
-* [Function Bodies](#function-bodies-section) section
-* [Data Segments](#data-segments-section) section
+* [Export](#export-section) section
+* [Start](#start-section) section
+* [Code](#code-section) section
+* [Data](#data-section) section
 * [Names](#names-section) section
 
 The end of the last present section must coincide with the last byte of the
 module. The shortest valid module is 8 bytes (`magic number`, `version`,
 followed by zero sections).
 
-### Signatures section
+### Type section
 
-ID: `signatures`
+ID: `types`
 
-The signatures section declares all function signatures that will be used in the module.
+The type section declares all function signatures that will be used in the module.
 
 | Field | Type | Description |
 | ----- |  ----- | ----- |
-| count | `varuint32` | count of signature entries to follow |
-| entries | `signature_entry*` | repeated signature entries as described below |
+| count | `varuint32` | count of type entries to follow |
+| entries | `type_entry*` | repeated type entries as described below |
 
-#### Signature entry
+#### Type entry
 | Field | Type | Description |
 | ----- |  ----- | ----- |
+| form | `uint8` | `0x40`, indicating a function type |
```

Comment:
On Wed, Apr 6, 2016 at 6:14 PM, rossberg-chromium <notifications@github.com>
wrote:

> In BinaryEncoding.md
> <https://github.com/WebAssembly/design/pull/640#discussion_r58734661>:
>
> >  | Field | Type | Description |
> >  | ----- |  ----- | ----- |
> > +| form | `uint8` | `0x40`, indicating a function type |
>
> I didn't understand this part of the OP: why would we want to add the
> complication of "inlining" certain structural constructors instead of just
> using a type id?
> E.g. to reduce the overhead of one-off uses of structural types. Or
> imagine you want to emit a more complicated nested struct, then you
> wouldn't need to separate out and name each level. Perhaps the other way
> round is a more compelling scenario: you may want to allow primitive
> constructors in type definitions. Or they'll start to mix when we introduce
> type import/exports some day. Also, it's hard to predict what other thing
> might come along and change the story (generics? who knows). I'm not
> suggesting that there currently is a concrete reason to join the spaces.
> But it doesn't seem completely unlikely to arise later, and given that the
> cost of keeping it an option is zero, why preclude it? Yes, if we can rule
> out the third case as I'm asking above then the
> encoding could be pretty simple: if positive, it's a pritimive, if
> negative, it's a (negated) type-id.
> I'd actually invert that scheme, to avoid the negation.
>

Oh, you went and mentioned generics did you!

If we do indeed end up having generics (let's please call them by their
right name, type parameters :-)), then we have type parameter declaration
sites at function declarations and struct type declarations, then type
arguments that occur at usage sites: calls and instantiations in the
bytecode.

One way to handle type parameters (i.e. declarations) is to have another
type constructor which essentially corresponds to the "forall" quantifier.

But that doesn't solve the type arguments problem. There will be a scoping
problem since type arguments are essentially type expressions with free
variables. That means that type arguments that occur in a function body
need to be considered to be within a "scope" that binds the type parameters.

But all of that is kind of yucky, let's please not have generics or type
parameters in WebAssembly. I think that it's entirely reasonable to expect
source language implementations to employ their own specialization/boxing
schemes to deal with a WebAssembly engine that has no concept of type
parameters.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/WebAssembly/design/pull/640/files/ec21463b44f46263fb81bdcb1f3cdf56a58bf93f#r58734661>
>